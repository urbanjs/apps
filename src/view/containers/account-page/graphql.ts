import { gql } from 'react-apollo';

export const updateUserMutation = gql`  
mutation updateUserPersonalInformation($userId: ID!, $data: UserPersonalInformationInput!) {
  updateUserPersonalInformation(userId: $userId, data: $data) {
    id
    firstName
    lastName
    birthDate
    phoneNumber
    birthPlace
    socialSecurityNumber
    taxNumber
    mothersMaidenName
  }
}`;

export const userQuery = gql`
query {
  user {
    id
    email
    displayName
    avatar
    personalInformation {
      id
      firstName
      lastName
      birthDate
      phoneNumber
      birthPlace
      socialSecurityNumber
      taxNumber
      mothersMaidenName
    }
  }
}`;
