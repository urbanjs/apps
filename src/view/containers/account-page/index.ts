export { AccountPageWithHOCs as AccountPage } from './account-page';
