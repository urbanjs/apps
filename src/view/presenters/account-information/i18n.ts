export default {
  edit: 'common.action.edit',
  title: 'component.accountInformation.title'
};
