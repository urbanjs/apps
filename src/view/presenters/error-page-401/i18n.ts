export default {
  'error': 'common.error',
  '401': 'common.error.401'
};
