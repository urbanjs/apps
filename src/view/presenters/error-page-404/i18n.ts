export default {
  'error': 'common.error',
  '404': 'common.error.404'
};
