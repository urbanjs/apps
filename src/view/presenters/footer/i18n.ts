export default {
  'copyright': 'component.footer.copyright'
};
