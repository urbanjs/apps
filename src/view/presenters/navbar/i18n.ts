export default {
  'link.login': 'common.action.login',
  'link.account': 'component.navbar.link.account'
};
