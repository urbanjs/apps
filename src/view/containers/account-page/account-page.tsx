import * as React from 'react';
import { connect } from 'react-redux';
import { Switch, Route, Redirect, withRouter, RouteComponentProps } from 'react-router-dom';
import { QueryProps as ApolloQueryProps, graphql } from 'react-apollo';
import {
  PATH_APP_404,
  PATH_APP_ACCOUNT,
  PATH_APP_ACCOUNT_EDIT
} from '../../../constants';
import { AccountInformation, AccountInformationEdit, ErrorPage401 } from '../../presenters';
import { RootState } from '../../../state/reducers';
import { updateUserMutation, userQuery } from './graphql';
import './account-page.css';

export type OwnProps = {};

export type StateProps = {
  serverOrigin: string;
};

export type User = {
  id: string;
  personalInformation: {}
};

export type UserQueryProps = {
  data: ApolloQueryProps & {
    user?: User
  }
};

export type UpdateUserMutationProps = {
  submit: (userId: string, data: object) => Promise<void>;
};

export type AccountPageProps =
  OwnProps
  & UserQueryProps
  & UpdateUserMutationProps
  & StateProps
  & RouteComponentProps<null>;

export class AccountPage extends React.Component<AccountPageProps> {
  props: AccountPageProps;

  render() {
    if (!this.props.data.loading && !this.props.data.user) {
      return <ErrorPage401/>;
    }

    const user = this.props.data.user as User;
    return (
      <div className="zv-account-page">
        <Switch>
          <Route
            path={PATH_APP_ACCOUNT_EDIT}
            exact={true}
            render={() => (
              <div className="m-2">
                <AccountInformationEdit
                  accountInformation={user.personalInformation}
                  onSave={async (data) => {
                    this.props.history.push(PATH_APP_ACCOUNT);
                    await this.props.submit(user.id, data.changes);
                  }}
                  onCancel={() =>
                    this.props.history.goBack()}
                />
              </div>
            )}
          />

          <Route
            path={PATH_APP_ACCOUNT}
            exact={true}
            render={() => (
              <div className="m-2">
                <AccountInformation
                  user={user}
                  onEdit={() =>
                    this.props.history.push(PATH_APP_ACCOUNT_EDIT)}
                />
              </div>
            )}
          />

          <Redirect to={PATH_APP_404}/>
        </Switch>
      </div>
    );
  }
}

export const AccountPageWithHOCs =
  graphql<UpdateUserMutationProps, OwnProps>(updateUserMutation, {
    props: ({mutate}) => ({
      submit: (userId: string, data: object) => {
        if (!mutate) {
          throw new Error('mutate does not exist');
        }

        return mutate({variables: {userId, data}});
      }
    }),
  })(
    graphql<UserQueryProps, OwnProps>(userQuery)(
      withRouter<OwnProps>(
        connect<StateProps>(
          (state: RootState): StateProps => ({
            serverOrigin: state.runtime.variables.serverOrigin
          })
        )(AccountPage))));
