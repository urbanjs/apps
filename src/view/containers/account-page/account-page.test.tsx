import * as React from 'react';
import { DocumentNode } from 'graphql';
import * as renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { IntlProvider } from 'react-intl';
import { MockedProvider as ApolloProvider } from 'react-apollo/test-utils';
import { Store, Provider as StoreProvider } from 'react-redux';
import { PATH_APP_ACCOUNT } from '../../../constants';
import { translations } from '../../../i18n';
import { createStore, RootState } from '../../../state';
import { AccountPageWithHOCs as AccountPage, OwnProps } from './account-page';
import { userQuery } from './graphql';

window.requestAnimationFrame = () => {
  throw new Error('requestAnimationFrame is not supported in Node');
};

describe('Account page', () => {
  let store: Store<RootState>;
  let apolloMocks: { request: { query: DocumentNode }, result: object }[];
  let currentPath: string;
  let i18nMessages: { [key: string]: string };

  beforeEach(() => {
    store = createStore({}, {
      platform: 'browser',
      devMode: false
    });

    apolloMocks = [];

    currentPath = PATH_APP_ACCOUNT;

    i18nMessages = translations.en;
  });

  const renderComponent = (props?: OwnProps) => {
    const component = (
      <MemoryRouter initialEntries={[currentPath]}>
        <StoreProvider store={store}>
          <ApolloProvider mocks={apolloMocks} removeTypename={true}>
            <IntlProvider locale="en" messages={i18nMessages}>
              <AccountPage {...props}/>
            </IntlProvider>
          </ApolloProvider>
        </StoreProvider>
      </MemoryRouter>
    );

    return renderer.create(component);
  };

  describe('if user is not authenticated', () => {
    beforeEach(() => {
      apolloMocks = [
        {
          request: {query: userQuery},
          result: {data: {user: null}}
        }
      ];
    });
    it('renders 401 error page ', async () => {
      const component = renderComponent();

      await new Promise(resolve => setTimeout(resolve, 5));
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  describe('if user is authenticated', () => {
    beforeEach(() => {
      apolloMocks = [
        {
          request: {query: userQuery},
          result: {
            data: {
              user: {
                id: 'id',
                email: 'email',
                displayName: 'displayName',
                avatar: 'avatar',
                personalInformation: {
                  id: 'id',
                  firstName: 'firstName',
                  lastName: 'lastName',
                  birthDate: 'birthDate',
                  phoneNumber: 'phoneNumber',
                  birthPlace: 'birthPlace',
                  socialSecurityNumber: 'socialSecurityNumber',
                  taxNumber: 'taxNumber',
                  mothersMaidenName: 'mothersMaidenName'
                }
              }
            }
          }
        }
      ];
    });

    it('renders page without error', async () => {
      const component = renderComponent();

      await new Promise(resolve => setTimeout(resolve, 5));
      expect(component.toJSON()).toMatchSnapshot();
    });
  });
});
