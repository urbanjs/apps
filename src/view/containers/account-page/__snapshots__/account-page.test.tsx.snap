// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account page if user is authenticated renders page without error 1`] = `
<div
  className="zv-account-page"
>
  <div
    className="m-2"
  >
    <div
      className="zv-account-information bg-faded p-4"
    >
      <a
        className="text-primary float-right btn btn-link m-3"
        onClick={[Function]}
      >
        <i
          className="fa fa-2x fa-pencil-square-o align-middle"
        />
        <br />
        <span>
          Edit
        </span>
      </a>
      <div
        className="p-4"
      >
        <i
          className="fa fa-3x fa-info align-bottom text-primary"
        />
        <h3
          className="d-inline-block ml-4 text-muted"
        >
          <span>
            Account
          </span>
        </h3>
      </div>
      <table
        className="table"
      >
        <tbody>
          <tr
            className="zv-group"
          >
            <td
              className="text-muted p-4"
            >
              <pre>
                {
                 "id": "id",
                 "email": "email",
                 "displayName": "displayName",
                 "avatar": "avatar",
                 "personalInformation": {
                  "id": "id",
                  "firstName": "firstName",
                  "lastName": "lastName",
                  "birthDate": "birthDate",
                  "phoneNumber": "phoneNumber",
                  "birthPlace": "birthPlace",
                  "socialSecurityNumber": "socialSecurityNumber",
                  "taxNumber": "taxNumber",
                  "mothersMaidenName": "mothersMaidenName"
                 }
                }
              </pre>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`Account page if user is not authenticated renders 401 error page  1`] = `
<div
  className="zv-error-page-401"
>
  <div
    className="cover"
  >
    <h1>
      Error
    </h1>
    <div
      className="lead"
    >
      <span>
        Forbidden
      </span>
    </div>
  </div>
</div>
`;
